[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/MaxMitre/Mate_MercadosFinancieros/blob/main/Semana2_Forwards.ipynb)

# Forwards

Hablemos de 2 conceptos fundamentales para esta clase.

- Ganancia/Pérdida del Forward
- Valuación del Forward


Primero, ¿cómo podríamos decidir cual sería el valor justo en un contrato tipo Forward?

In [ ]:
# !pip install yfinance

In [ ]:
import numpy as np
import pandas as pd

import yfinance as yf

Recordemos el término de valor presente o dar valor a futuro (son muy parecidos)

In [ ]:
data = yf.download(  # or pdr.get_data_yahoo(...
        # tickers list or string as well
        tickers = "GOOG NVDA",

        # use "period" instead of start/end
        # valid periods: 1d,5d,1mo,3mo,6mo,1y,2y,5y,10y,ytd,max
        # (optional, default is '1mo')
        period = "1y",

        # fetch data by interval (including intraday if period < 60 days)
        # valid intervals: 1m,2m,5m,15m,30m,60m,90m,1h,1d,5d,1wk,1mo,3mo
        # (optional, default is '1d')
        interval = "1d",

        # group by ticker (to access via data['SPY'])
        # (optional, default is 'column')
        # group_by = 'ticker',
    ).loc[:, 'Adj Close']
data

[*********************100%%**********************]  2 of 2 completed


Ticker            GOOG        NVDA
Date                              
2023-07-20  119.394135   45.506374
2023-07-21  120.173248   44.295734
2023-07-24  121.741463   44.598648
2023-07-25  122.650429   45.665329
2023-07-26  129.512619   45.438400
...                ...         ...
2024-07-15  188.190002  128.440002
2024-07-16  185.500000  126.360001
2024-07-17  182.619995  117.989998
2024-07-18  179.220001  121.089996
2024-07-19  179.389999  117.930000

[252 rows x 2 columns]

* ¿Es necesario aclarar la diferencia entre una Series y un DataFrame en pandas?

In [ ]:
data[['GOOG']].shape

(252, 1)

In [ ]:
data['GOOG'].shape

(252,)

# Precio justo

In [ ]:
spot_price = data['NVDA'][-1]
spot_price

117.93000030517578

In [ ]:
# Precio justo, suponiendo que la tasa libre de riesgo es constante
r = 0.052
T = 90/365
forward_price = spot_price * (1 + r*T)
forward_price

119.44208907621201

In [ ]:
# Precio justo cuando se toma como proceso continuo
r = 0.052
T = 90/365
forward_price = spot_price * np.exp(r*T)
forward_price

119.45182457996158

# Ejercicio

Haga el código del siguiente escenario:
Ves el precio actual de NVIDIA (aprox 117), calculas el valor futuro de modo sencillo y decides entrar en un contrato Forward en Largo por 1000 acciones que vencerá en 30 días (te comprometes a comprar). 10 días después, el activo subió a 130, y decides entrar en un contrato de tipo Forward en Corto y lo haces a un precio justo.

¿Que ganancia obtendrías?

¿Si decides ejecutar tus contratos, que ganancia tendrías al traerla a valor presente?

In [ ]:
# Espacio para el ejercicio (pueden agregar mas celdas de código)

In [ ]:
ganancia_final = 11936.381063382925

* Evento de Gamestop

# Valuación de un Forward

$$ V_t = S_t - F e^{r(T - t)} $$

In [ ]:
def forward_contract_value(spot_price, forward_price, risk_free_rate, time_to_maturity, current_time):
    # Calcula el VALOR del contrato
    value = spot_price - forward_price * np.exp(risk_free_rate * (time_to_maturity - current_time))
    return value

In [ ]:
# Parametros
spot_price = 100
forward_price = 105
risk_free_rate = 0.05
time_to_maturity = 365/365 # 1
current_time = 1/365

In [ ]:
value = forward_contract_value(spot_price, forward_price, risk_free_rate, time_to_maturity, current_time)

print(f"El valor del contrato es: {value}")

El valor del contrato es: -10.368345132502597


$$ f(s_t, t) = v_t $$

$$ f^{'} (s_t, t)= v^{'}_t$$

# Paréntesis cultural

Tratar con datos tipo fecha en python (se pueden tomar del móduo "datetime" o directamente en pandas con auxiliares)

In [ ]:
fecha = pd.to_datetime('07-08-2024')
fecha

Timestamp('2024-07-08 00:00:00')

In [ ]:
fecha.month

7

In [ ]:
# ¿Como corregir que lea primero el día y luego el mes (diferente a USA)?
fecha = pd.to_datetime('07-08-2024', dayfirst=True)

In [ ]:
fecha.month

8

In [ ]:
dif = pd.to_datetime('07/08/2024', dayfirst=True) - pd.to_datetime('01/01/2024', dayfirst=True)
dif

Timedelta('219 days 00:00:00')

In [ ]:
dif.days

219

In [ ]:
# Como ver atributos de mi objeto
dir(dif)

['__abs__',
 '__add__',
 '__array_priority__',
 '__bool__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__divmod__',
 '__doc__',
 '__eq__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__mod__',
 '__module__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__pos__',
 '__pyx_vtable__',
 '__radd__',
 '__rdivmod__',
 '__reduce__',
 '__reduce_cython__',
 '__reduce_ex__',
 '__repr__',
 '__rfloordiv__',
 '__rmod__',
 '__rmul__',
 '__rsub__',
 '__rtruediv__',
 '__setattr__',
 '__setstate__',
 '__setstate_cython__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__truediv__',
 '__weakref__',
 '_creso',
 '_d',
 '_from_value_and_reso',
 '_h',
 '_is_populated',
 '_m',
 '_maybe_cast_to_matching_resos',
 '_microseconds',
 '_ms',
 '_ns',
 '_repr_base',
 '_req_any_kwargs_new',
 '_round',
 '_s',
 '_seconds',
 '_unit',
 '_us',
 '_value',
 'as_unit',
 'asm8',
 'ceil',
 'com

# Ejercicio 2

1.- Cree una tabla en pandas que tenga 6 columnas, 5 van a ser los parámetros utilizados para ver el valor de un forward y la sexta será el valor del contrato.

2.- Visualice (gráfico) como se ve el cambio del precio del Forward que se basa en NVIDIA al avanzar en el tiempo

In [ ]:
# Espacio para ejercicio 2


# Redes Neuronales

In [ ]:
from tensorflow import keras
import tensorflow as tf

model = keras.Sequential(
    [
        keras.layers.Dense(64, activation="relu", input_shape=(5,)),
        keras.layers.Dense(64, activation="relu"),
        keras.layers.Dense(64, activation="relu"),
        keras.layers.Dense(1, activation="sigmoid")
    ]
)
model.summary()

In [ ]:
model.compile(
    optimizer=keras.optimizers.Adam(0.01), loss="mse",
)

history = model.fit(
    "train_features",
    "train_targets",
    batch_size=20, # Cantidad de datos que optimizan pesos por cada pasada, "batch size"
    epochs=30, # Cantidad de pasadas que dan todos los datos, para entrenamiento
    verbose=2,
)

¿Podríamos hacer una red que haga predicciones recibiendo solo el tiempo como parámetro?